flow-server/src/main/java/com/vaadin/flow/server/auth/ViewAccessChecker.java

```diff
@@ -0,0 +1,176 @@
+        throwIfEnabled();
+        this.loginView = loginView;
+    }
+
+    /**
+     * Sets the Fusion login view to use.
+     * <p>
+     * Note that the access checker needs to be separately enabled using
+     * {@link #setEnabled(boolean)}
+     * 
+     * @param loginUrl
+     *            the Fusion view to use as login view
+     */
+    public void setLoginView(String loginUrl) {
+        throwIfEnabled();
+        this.loginUrl = loginUrl;
+    }
+
+    private void throwIfEnabled() {
+        if (this.loginUrl != null) {
+            throw new IllegalStateException(
+                    "Already using " + this.loginUrl + " as the login view");
+        }
+        if (this.loginView != null) {
+            throw new IllegalStateException("Already using "
+                    + this.loginView.getName() + " as the login view");
+        }
+    }
+
+    @Override
+    public void beforeEnter(BeforeEnterEvent beforeEnterEvent) {
+        if (!isEnabled()) {
+            return;
+        }
+        Class<?> targetView = beforeEnterEvent.getNavigationTarget();
+
+        VaadinServletRequest vaadinServletRequest = VaadinServletRequest
+                .getCurrent();
+        if (vaadinServletRequest == null) {
+            // This is in a background thread and we cannot access the request
+            // to check access
+            getLogger().warn("Preventing navigation to " + targetView.getName()
+                    + " because no HTTP request is available for checking access.");
+            beforeEnterEvent.rerouteToError(NotFoundException.class);
+            return;
+        }
+
+        HttpServletRequest httpServletRequest = vaadinServletRequest
+                .getHttpServletRequest();
+        getLogger().debug("Checking access for view {}", targetView.getName());
+        if (loginView != null && targetView == loginView) {
+            getLogger().debug("Allowing access for login view {}",
+                    targetView.getName());
+            return;
+        }
+        boolean ok = accessAnnotationChecker.hasAccess(targetView,
+                httpServletRequest);
+
+        if (ok) {
+            getLogger().debug("Allowed access to view {}",
+                    targetView.getName());
+            return;
+        } else {
+            getLogger().debug("Denied access to view {}", targetView.getName());
+            if (httpServletRequest.getUserPrincipal() == null) {
+                httpServletRequest.getSession()
+                        .setAttribute(SESSION_STORED_REDIRECT, beforeEnterEvent.getLocation().getPathWithQueryParameters());
+                if (loginView != null) {
+                    beforeEnterEvent.forwardTo(loginView);
+                } else {
+                    beforeEnterEvent.getUI().getPage().setLocation(loginUrl);
+                }
+            } else {
+                beforeEnterEvent.rerouteToError(NotFoundException.class);
+            }
+        }
```

Comment:
I suggest to reduce the indentation of these lines by ommitting `} else {` block after `if (ok) { return }`:

```suggestion
        getLogger().debug("Denied access to view {}", targetView.getName());
        if (httpServletRequest.getUserPrincipal() == null) {
            httpServletRequest.getSession()
                    .setAttribute(SESSION_STORED_REDIRECT, beforeEnterEvent.getLocation().getPathWithQueryParameters());
            if (loginView != null) {
                beforeEnterEvent.forwardTo(loginView);
            } else {
                beforeEnterEvent.getUI().getPage().setLocation(loginUrl);
            }
        } else {
            beforeEnterEvent.rerouteToError(NotFoundException.class);
        }
```